File: samples/java/shorturls/app/controllers/Application.java
package controllers;

import models.ShortURL;
import play.libs.F;
import play.mvc.Controller;
import play.mvc.Result;

public class Application extends Controller {

    public static Result index() {
        return ok(views.html.index.render());
    }

    public static Result goTo(String id) {
        return async(
            ShortURL.findById(id).map(new F.Function<ShortURL, Result>() {
                @Override
                public Result apply(ShortURL shortURL) throws Throwable {
                    if (shortURL == null) {
                        return notFound();
                    }
                    return redirect(shortURL.originalUrl);
                }
            })
        );
    }
}

File: samples/scala/shorturls/app/controllers/Application.scala
package controllers

import play.api.mvc._
import models.ShortURLs
import models.ShortURLs._

object Application extends Controller {
  
  def index = Action {
    Ok(views.html.index())
  }

  def goTo(id: String) = Action.async {
    ShortURLs.findById(id).map { maybe =>
      maybe.map( url => Redirect(url.originalUrl) ).getOrElse(NotFound)
    }
  }
}